CRYPTOGRAPHY.md

```diff
@@ -2,120 +2,526 @@
 
+
+## Snapshots
+A snapshot is essentially a virtual filesystem that describes the layout of directories entrie, file entries and file data among a few other things.
+Each node of the virtual filesystem is either a structure that points to other structures or to raw chunks of data.
+
+All these nodes are stored as blobs within packfiles,
+and pointers within the virtual filesystems are represented as checksums that can be looked up into packfiles to fetch related blobs.
+
+Each snapshot gets assigned a random 256-bits identifier that points to a virtual filesystem root,
+and browsing the snapshot consists in resolving children nodes from there by performing the proper packfile lookups and fetches.
+
+```
+$ plakar ls e8b104b9:/private/etc     
+repository passphrase: 
+2024-12-07T08:11:55Z drwxr-xr-x                      160 B uucp
+
+$ ./plakar ls e8b104b9:/private/etc/uucp
+repository passphrase: 
+2024-12-07T08:11:55Z -r--r--r--                      133 B passwd
+2024-12-07T08:11:55Z -r--r--r--                      141 B port
+2024-12-07T08:11:55Z -r--r--r--                     2.4 kB sys
+$
 ```
-[encrypted blob]
-[encrypted blob]
-[encrypted blob]
+
+### The backup process
+
+```
+$ plakar -trace repository,snapshot,packer backup /private/etc/uucp
 [...]
-[encrypted blob]
-[encrypted blob]
-[encrypted packfile index]
-[encrypted packfile footer]
 ```
 
-Each blob is encrypted individually so that they can be accessed and decrypted individually as long as their location and length is known.
-The index and footer are also encrypted individually for the same reason.
+An empty state is created to hold changes to the current local cache.
 
+As backup progresses, chunks of data are produced from various sources:
+file content, filesystem structure, plakar internal structures, ...
 
-#### Packfile index
+Each time a chunk is produced,
+a checksum of the data is computed for internal purposes and recording within the snapshot itself:
 
-The packfile index stores the following record for each blob:
 ```
-[type][cleartext checksum][offset][length]
+dataChecksum = HASH(data)
 ```
 
-#### Packfile footer
+To determnine if the chunk already exists in the repository or if it needs to be pushed,
+a blob identifier is computed by performing an HMAC of the checksum with the master key:
 
-The packfile footer stores a few metadata such as creation data, number of items, offset of the index and the checksum of the packfile index as a fast integrity check.
+```
+blobId = HMAC(dataChecksum, masterKey)
+```
 
+That blob identifier is then used as a key to query local cache:
 
-## Backup process
+```
+found = localCache.BlobExists(blobType, blobId)
+```
 
-A snapshot header is created with a random 256-bits identifier.
+If blob is found, then a blob with the same checksum already exists in the repository and there is no need to push it again.
```

Comment:
if a *blob id* is found?  otherwise it seems we're searching the blob themselves

Reply:
solved